File: starlark/docgen/ast/parser_test.go
// Copyright 2019 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package ast

import (
	"fmt"
	"io"
	"strings"
	"testing"

	"go.starlark.net/syntax"

	"go.chromium.org/luci/common/testing/ftt"
	"go.chromium.org/luci/common/testing/truth/assert"
	"go.chromium.org/luci/common/testing/truth/should"
)

const goodInput = `
# Copyright comment or something.

"""Module doc string.

Multiline.
"""

# Load comment.
load('@stdlib//another.star', 'ext1', ext2='ext_name')

# Skipped comment.

# Function comment.
#
#   With indent.
#
# More.
def func1(*, a, b, c=None, **kwargs):
  """Doc string.

  Multiline.
  """
  body

# No docstring
def func2():
  pass

# Weird doc string: number instead of a string.
def func3():
  42

const_int = 123
const_str = 'str'

ellipsis = 1 + 2

alias1 = func1
alias2 = ext2.deeper.deep

decl = namespace.declare(
    skipped,
    arg1 = 123,
    arg2 = 1 + 1,
    arg3 = func1,
    arg4 = ext2.deeper.deep,
    arg5 = nested(nested_arg = 1),
)

# Struct comment.
struct_stuff = struct(
    const = 123,
    stuff = 1+2+3,
    # Key comment 1.
    key1 = func1,
    key2 = ext2.deeper.deep,
    # Nested namespace.
    nested = struct(key1=v1, key2=v2),

    **skipped
)

skipped, skipped = a, b
skipped += 123
`

const expectedDumpOutput = `mod.star = module
  ext1 -> ext1 in @stdlib//another.star
  ext2 -> ext_name in @stdlib//another.star
  func1 = func
  func2 = func
  func3 = func
  const_int = 123
  const_str = str
  ellipsis = ...
  alias1 = func1
  alias2 = ext2.deeper.deep
  decl = namespace.declare(...)
    arg1 = 123
    arg2 = ...
    arg3 = func1
    arg4 = ext2.deeper.deep
    arg5 = nested(...)
      nested_arg = 1
  struct_stuff = namespace
    const = 123
    stuff = ...
    key1 = func1
    key2 = ext2.deeper.deep
    nested = namespace
      key1 = v1
      key2 = v2
`

func TestParseModule(t *testing.T) {
	t.Parallel()

	ftt.Run("Works", t, func(t *ftt.Test) {
		mod, err := ParseModule(&syntax.FileOptions{}, "mod.star", goodInput, func(s string) (string, error) { return s, nil })
		assert.Loosely(t, err, should.BeNil)
		buf := strings.Builder{}
		dumpTree(mod, &buf, "")
		assert.Loosely(t, buf.String(), should.Equal(expectedDumpOutput))

		t.Run("Docstrings extraction", func(t *ftt.Test) {
			assert.Loosely(t, mod.Doc(), should.Equal("Module doc string.\n\nMultiline.\n"))
			assert.Loosely(t,
				nodeByName(mod, "func1").Doc(), should.Equal(
					"Doc string.\n\n  Multiline.\n  ",
				))
		})

		t.Run("Spans extraction", func(t *ftt.Test) {
			l, r := nodeByName(mod, "func1").Span()
			assert.Loosely(t, l.Filename(), should.Equal("mod.star"))
			assert.Loosely(t, r.Filename(), should.Equal("mod.star"))

			// Spans the entire function definition.
			assert.Loosely(t, extractSpan(goodInput, l, r), should.Equal(`def func1(*, a, b, c=None, **kwargs):
  """Doc string.

  Multiline.
  """
  body`))
		})

		t.Run("Comments extraction", func(t *ftt.Test) {
			assert.Loosely(t, nodeByName(mod, "func1").Comments(), should.Equal(`Function comment.

  With indent.

More.`))

			strct := nodeByName(mod, "struct_stuff").(*Namespace)
			assert.Loosely(t, strct.Comments(), should.Equal("Struct comment."))

			// Individual struct entries are annotated with comments too.
			assert.Loosely(t, nodeByName(strct, "key1").Comments(), should.Equal("Key comment 1."))

			// Nested structs are also annotated.
			nested := nodeByName(strct, "nested").(*Namespace)
			assert.Loosely(t, nested.Comments(), should.Equal("Nested namespace."))

			// Keys do not pick up comments not intended for them.
			assert.Loosely(t, nodeByName(nested, "key1").Comments(), should.BeEmpty)

			// Top module comment is not extracted currently, it is relatively hard
			// to do. We have a docstring though, so it's not a big deal.
			assert.Loosely(t, mod.Comments(), should.BeEmpty)
		})
	})
}

func dumpTree(nd Node, w io.Writer, indent string) {
	// recurseInto is used to visit Namespace and Module.
	recurseInto := func(nodes []Node) {
		if len(nodes) != 0 {
			for _, n := range nodes {
				dumpTree(n, w, indent+"  ")
			}
		} else {
			fmt.Fprintf(w, "%s  <empty>\n", indent)
		}
	}

	switch n := nd.(type) {
	case *Var:
		fmt.Fprintf(w, "%s%s = %v\n", indent, n.name, n.Value)
	case *Function:
		fmt.Fprintf(w, "%s%s = func\n", indent, n.name)
	case *Reference:
		fmt.Fprintf(w, "%s%s = %s\n", indent, n.name, strings.Join(n.Path, "."))
	case *ExternalReference:
		fmt.Fprintf(w, "%s%s -> %s in %s\n", indent, n.name, n.ExternalName, n.Module)
	case *Invocation:
		fmt.Fprintf(w, "%s%s = %s(...)\n", indent, n.name, strings.Join(n.Func, "."))
		recurseInto(n.Args)
	case *Namespace:
		fmt.Fprintf(w, "%s%s = namespace\n", indent, n.name)
		recurseInto(n.Nodes)
	case *Module:
		fmt.Fprintf(w, "%s%s = module\n", indent, n.name)
		recurseInto(n.Nodes)
	default:
		panic(fmt.Sprintf("unknown node kind %T", nd))
	}
}

func nodeByName(n EnumerableNode, name string) Node {
	for _, node := range n.EnumNodes() {
		if node.Name() == name {
			return node
		}
	}
	return nil
}

func extractSpan(body string, start, end syntax.Position) string {
	lines := strings.Split(body, "\n")

	// Note: sloppy, but good enough for the test. Also note that Line and Col
	// are 1-based indexes.
	var out []string
	out = append(out, lines[start.Line-1][start.Col-1:])
	out = append(out, lines[start.Line:end.Line-1]...)
	out = append(out, lines[end.Line-1][:end.Col-1])

	return strings.Join(out, "\n")
}

File: milo/ui/src/fleet/components/filter_dropdown/parser/parser.test.ts
// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { BLANK_VALUE } from '@/fleet/constants/filters';

import { parseFilters, stringifyFilters } from './parser';

describe('stringifyFilters', () => {
  it('should stringify a single filter', () => {
    expect(stringifyFilters({ key: ['value'] })).toBe('key = "value"');
  });

  it('should stringify a multi-value filter', () => {
    expect(stringifyFilters({ key: ['v1', 'v2'] })).toBe(
      'key = ("v1" OR "v2")',
    );
  });

  it('should stringify a blank filter', () => {
    expect(stringifyFilters({ key: [BLANK_VALUE] })).toBe('NOT key');
  });

  it('should stringify a blank and single value filter', () => {
    expect(stringifyFilters({ key: [BLANK_VALUE, 'v1'] })).toBe(
      '(NOT key OR key = "v1")',
    );
  });

  it('should stringify a blank and multi-value filter', () => {
    expect(stringifyFilters({ key: [BLANK_VALUE, 'v1', 'v2'] })).toBe(
      '(NOT key OR key = ("v1" OR "v2"))',
    );
  });

  it('should stringify multiple filters', () => {
    expect(
      stringifyFilters({
        key1: ['v1'],
        key2: [BLANK_VALUE],
        key3: ['v2', 'v3'],
        key4: [BLANK_VALUE, 'v4'],
      }),
    ).toBe(
      'key1 = "v1" NOT key2 key3 = ("v2" OR "v3") (NOT key4 OR key4 = "v4")',
    );
  });
});

describe('parseFilters', () => {
  it('should parse a single filter', () => {
    expect(parseFilters('key = "value"').filters).toEqual({ key: ['value'] });
  });

  it('should parse a filters with key like labels.key quoting the key part', () => {
    expect(parseFilters('labels.key = "value"').filters).toEqual({
      'labels."key"': ['value'],
    });
  });

  it('should parse a multi-value filter', () => {
    expect(parseFilters('key = ("v1" OR "v2")').filters).toEqual({
      key: ['v1', 'v2'],
    });
  });

  it('should parse a blank filter', () => {
    expect(parseFilters('NOT key').filters).toEqual({ key: [BLANK_VALUE] });
  });

  it('should parse a blank and single value filter', () => {
    expect(parseFilters('(NOT key OR key = "v1")').filters).toEqual({
      key: [BLANK_VALUE, 'v1'],
    });
  });

  it('should parse a blank and multi-value filter', () => {
    expect(parseFilters('(NOT key OR key = ("v1" OR "v2"))').filters).toEqual({
      key: [BLANK_VALUE, 'v1', 'v2'],
    });
  });

  it('should parse multiple filters', () => {
    const result = parseFilters(
      'key1 = "v1" NOT key2 key3 = ("v2" OR "v3") (NOT key4 OR key4 = "v4")',
    );
    expect(result.error).toBeUndefined();
    expect(result.filters).toEqual({
      key1: ['v1'],
      key2: [BLANK_VALUE],
      key3: ['v2', 'v3'],
      key4: [BLANK_VALUE, 'v4'],
    });
  });

  it('should parse labels with weird characters', () => {
    const result = parseFilters('(NOT label OR label="this:\\"isNOT(weird)")');
    expect(result.error).toBeUndefined();
    expect(result.filters).toEqual({
      label: [BLANK_VALUE, 'this:\\"isNOT(weird)'],
    });
  });

  it('should return an error for invalid syntax', () => {
    expect(parseFilters('key = (v1 OR v2').error).toBeInstanceOf(Error);
    expect(parseFilters('(NOT key OR key = "v1"').error).toBeInstanceOf(Error);
    expect(parseFilters('key = ').error).toBeInstanceOf(Error);
  });
});
